File: server/response_payload.py
from sys import getsizeof

class ResponsePayload:
  def __init__(self, client_id):
    self.client_id = client_id

  def get_client_id(self):
    return self.client_id
  
  def get_size(self):
    attributes = [attr for attr in dir(self) if not attr.startswith('__') and not callable(getattr(self, attr))]
    total_size = sum(getsizeof(getattr(self, attr)) for attr in attributes)
    return total_size
  
class ResponseEmptyPayload(ResponsePayload):
  def get_size(self):
    return 0
class ResponseKeyPayload(ResponsePayload):
  def __init__(self, client_id, encrypted_key):
    super().__init__(client_id)
    self.encrypted_key = encrypted_key

  def get_encrypted_key(self):
    return self.encrypted_key
  
class ResponseReceiveFilePayload(ResponsePayload):
  def __init__(self, client_id, content_size, file_name, cksum):
    super().__init__(client_id)
    self.content_size = content_size
    self.file_name = file_name
    self.cksum = cksum

  def get_content_size(self):
    return self.content_size
  
  def get_file_name(self):
    return self.file_name
  
  def get_cksum(self):
    return self.cksum

File: client/response_payload.cpp
#include <cstring>
#include "response_payload.h"

ResponseEmptyPayload::ResponseEmptyPayload() : size(0) {}

uint32_t ResponseEmptyPayload::get_size() {
  return size;
}

ResponsePayload::ResponsePayload(std::string client_id) {
  strncpy(this->client_id, client_id.c_str(), 16);
}

std::string ResponsePayload::get_client_id() {
  return std::string(this->client_id);
}

ResponseKeyPayload::ResponseKeyPayload(std::string client_id, std::string encrypted_key) 
  : ResponsePayload(client_id) 
  , encrypted_key(encrypted_key) {}

std::string ResponseKeyPayload::get_encrypted_key() {
  return encrypted_key;
}

ResponseReceiveFilePayload::ResponseReceiveFilePayload(std::string client_id, uint32_t content_size, std::string file_name, uint32_t cksum) 
  : ResponsePayload(client_id)
  , content_size(content_size)
  , cksum(cksum) {
  strncpy(this->file_name, file_name.c_str(), 255);
}

uint32_t ResponseReceiveFilePayload::get_content_size() {
  return content_size;
}

std::string ResponseReceiveFilePayload::get_file_name() {
  return std::string(this->file_name);
}

uint32_t ResponseReceiveFilePayload::get_cksum() {
  return cksum;
}

